File: sdk/dart/lib/src/resources/slugs.dart
// Copyright 2026 Oddbit (https://oddbit.id)
// SPDX-License-Identifier: Apache-2.0

import '../base_client.dart';
import '../models.dart';

/// Methods for slug lookup and management endpoints.
class SlugsResource {
  SlugsResource(this._http);

  final ShrtnrBaseClient _http;

  /// Look up a link by its slug.
  Future<Link> lookup(String slug) async {
    final encoded = Uri.encodeComponent(slug);
    final json = await _http.requestJson('GET', '/_/api/slugs/$encoded');
    return Link.fromJson(json! as Map<String, dynamic>);
  }

  /// Add a custom slug to a link.
  Future<Slug> add(int linkId, String slug) async {
    final json = await _http.requestJson(
      'POST',
      '/_/api/links/$linkId/slugs',
      body: <String, dynamic>{'slug': slug},
    );
    return Slug.fromJson(json! as Map<String, dynamic>);
  }

  /// Disable a specific slug on a link.
  Future<Slug> disable(int linkId, String slug) async {
    final encoded = Uri.encodeComponent(slug);
    final json = await _http.requestJson(
      'POST',
      '/_/api/links/$linkId/slugs/$encoded/disable',
    );
    return Slug.fromJson(json! as Map<String, dynamic>);
  }

  /// Re-enable a disabled slug on a link.
  Future<Slug> enable(int linkId, String slug) async {
    final encoded = Uri.encodeComponent(slug);
    final json = await _http.requestJson(
      'POST',
      '/_/api/links/$linkId/slugs/$encoded/enable',
    );
    return Slug.fromJson(json! as Map<String, dynamic>);
  }

  /// Remove a custom slug from a link.
  Future<RemovedResult> remove(int linkId, String slug) async {
    final encoded = Uri.encodeComponent(slug);
    final json = await _http.requestJson(
      'DELETE',
      '/_/api/links/$linkId/slugs/$encoded',
    );
    return RemovedResult.fromJson(json! as Map<String, dynamic>);
  }
}

File: sdk/typescript/src/resources/slugs.ts
// Copyright 2026 Oddbit (https://oddbit.id)
// SPDX-License-Identifier: Apache-2.0

import { HttpClient } from "../internal/http";
import { Link, RemovedResult, Slug } from "../models";

export class SlugsResource {
  constructor(private readonly http: HttpClient) {}

  /** Look up a link by its slug. */
  lookup(slug: string): Promise<Link> {
    return this.http.request("GET", `/_/api/slugs/${encodeURIComponent(slug)}`);
  }

  /** Add a custom slug to a link. */
  add(linkId: number, slug: string): Promise<Slug> {
    return this.http.request("POST", `/_/api/links/${linkId}/slugs`, {
      body: { slug },
    });
  }

  /** Disable a specific slug on a link. */
  disable(linkId: number, slug: string): Promise<Slug> {
    return this.http.request(
      "POST",
      `/_/api/links/${linkId}/slugs/${encodeURIComponent(slug)}/disable`,
    );
  }

  /** Re-enable a disabled slug on a link. */
  enable(linkId: number, slug: string): Promise<Slug> {
    return this.http.request(
      "POST",
      `/_/api/links/${linkId}/slugs/${encodeURIComponent(slug)}/enable`,
    );
  }

  /** Remove a custom slug from a link. */
  remove(linkId: number, slug: string): Promise<RemovedResult> {
    return this.http.request(
      "DELETE",
      `/_/api/links/${linkId}/slugs/${encodeURIComponent(slug)}`,
    );
  }
}
